controller/array_action/array_mediator_ds8k.py

```diff
@@ -370,17 +391,131 @@ def unmap_volume(self, volume_id, host_name):
+                logger.info("{} not found".format(fc_id))
+                if not_exist_err:
+                    raise ex
+        except Exception as ex:
+            logger.exception(ex)
+            raise ex
+
     def get_snapshot(self, snapshot_name):
-        # TODO: CSI-1339
-        pass
+        logger.debug("Get snapshot : {}".format(snapshot_name))
+        candidate_api_volume = self._get_api_volume_by_name(snapshot_name)
+        if not candidate_api_volume:
+            return None
+        if not candidate_api_volume.flashcopy:
+            logger.error("FlashCopy Mapping not found for target volume: {}".format(snapshot_name))
+            raise array_errors.SnapshotNameBelongsToVolumeError(candidate_api_volume.name,
+                                                                self.service_address)
+        fcmap = self._get_flashcopy(candidate_api_volume.flashcopy[0].id)
+        source_volume_name = self.get_volume_name(fcmap.source_volume['id'])
+        return self._generate_snapshot_response(candidate_api_volume, source_volume_name)
+
+    def _create_similar_volume(self, target_volume_name, source_volume_name):
+        logger.info(
+            "creating target api volume '{0}' from source volume '{1}'".format(target_volume_name,
+                                                                               source_volume_name))
+        source_api_volume = self._get_api_volume_by_name(source_volume_name)
+        if not source_volume_name:
+            raise array_errors.VolumeNotFoundError(source_volume_name)
+        capabilities = {config.CAPABILITIES_SPACEEFFICIENCY: None}
+        size_in_bytes = int(source_api_volume.cap)
+        pool = source_api_volume.pool
+        return self.create_volume(target_volume_name, size_in_bytes, capabilities, pool)
+
+    def _create_flashcopy(self, source_volume, target_volume):
+        logger.info(
+            "creating FlashCopy Mapping from '{0}' to '{1}'".format(source_volume.volume_name,
+                                                                    target_volume.volume_name))
+        source_volume_id = get_volume_id_from_scsi_identifier(source_volume.id)
+        target_volume_id = get_volume_id_from_scsi_identifier(target_volume.id)
+        volume_pair = flashcopy_request_volume_pair_parser(source_volume_id, target_volume_id)
+        try:
+            api_flashcopy = self.client.create_flashcopy(volume_pairs=volume_pair,
+                                                         options=["persistent"])
+        except (exceptions.ClientError, exceptions.ClientException) as ex:
+            if ERROR_CODE_ALREADY_FLASHCOPY in str(ex.message).upper():
+                raise array_errors.SnapshotAlreadyExists
+            elif ERROR_CODE_VOLUME_NOT_FOUND_OR_ALREADY_PART_OF_CS_RELATIONSHIP in str(
+                    ex.message).upper():
+                raise array_errors.VolumeNotFoundError
```

Comment:
I don't think this line works at runtime.
please add a unit test to cover it (i.e. add a test that uses `00000013`)

Reply:
done

Reply:
I think `VolumeNotFoundError` in `create_snapshot` context refers to the source volume, and since `_create_snapshot` ensures its existence before reaching here, this should probably be an idempotent case.
what do you think?

Reply:
as the name of the error suggest, it raise not only on volume not found, because I dont have the error doc yet I dont want to remove it now. its better to resolve this in CSI-1376

Reply:
I attach it to the ticket

Reply:
could you add the ticket number here? CSI-1376 seems closed